### PR TITLE
Feature/df 1245 upload component + BU create/edit

### PR DESCRIPTION
**Pull request by 99999989**

### CODE REVIEW

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document-modal/application-document-modal.component.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+})
+export class ApplicationDocumentModalComponent implements OnInit {
+
+  readonly ACCEPTED_FILE_TYPES = 'application/pdf';
+
+  readonly MAX_FILE_SIZE = 5 * 1024 * 1024; // 5 MB
+
+  @Input() applicationDocument: ApplicationDocument;
+
+  @Input() invalidatedDocumentTypes: ApplicationDocumentType[];
```
Comment (pavax) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document-modal/application-document-modal.component.ts]: rename to invalidDocumentTypes

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.component.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@Component({
+  selector: 'alv-application-document',
+  templateUrl: './application-document.component.html',
+  styleUrls: ['./application-document.component.scss']
+})
+export class ApplicationDocumentComponent {
+
+  @Input() applicationDocument: ApplicationDocument;
+
+  constructor(private modalService: ModalService,
```
Comment (pavax) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.component.ts]: delete modalService

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.model.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/**
+ * A "view-model" for the Work-Effort Component that has all of our business logic in it to
+ * avoid duplication and calculations in functions. Delegate pattern.
+ */
+import { ApplicationDocument } from '../../../../shared/backend-services/application-documents/application-documents.types';
+
+
+export class ApplicationDocumentModel {
```
Comment (pavax) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.model.ts]: delete ?

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  applySortByDocumentType() {
+    this.applicationDocuments = this.applicationDocumentsSortingService.sortByDocumentType(this.applicationDocuments);
+  }
+
+  applySortByFilename() {
+    this.applicationDocuments = this.applicationDocumentsSortingService.sortByFilename(this.applicationDocuments);
+  }
+
+  private getInvalidatedDocumentTypes(excludedDocumentType: ApplicationDocumentType): ApplicationDocumentType[] {
```
Comment (pavax) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: rename to calculateInvalidDocumenTypes

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    return Object.values(ApplicationDocumentType).filter(
```
Comment (pavax) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: maybe use a reduce function...?

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-sorting.service.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  sortByDate(applicationDocuments: ApplicationDocument[]): ApplicationDocument[] {
+    return applicationDocuments.sort((a, b) => {
+      return new Date(a.updatedAt).getTime() - new Date(b.updatedAt).getTime();
+    });
+  }
+
+  sortByDocumentType(applicationDocuments: ApplicationDocument[]): ApplicationDocument[] {
+    return applicationDocuments.sort((a, b) => {
+      if (a.documentType === b.documentType) {
+        return new Date(a.updatedAt).getTime() - new Date(b.updatedAt).getTime();
```
Comment (pavax) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-sorting.service.ts]: can we re-use the line: 
```
new Date(a.updatedAt).getTime() - new Date(b.updatedAt).getTime();
```

File: alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.repository.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  ApplicationDocument, ApplicationDocumentType,
+  CreateApplicationDocument,
+} from './application-documents.types';
+
+@Injectable({ providedIn: 'root' })
+export class ApplicationDocumentsRepository {
+
+  private readonly resourceUrl = '/onlineform-service/api/bu';
+
+  private readonly searchUrl = `${this.resourceUrl}/_search`;
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.repository.ts]: delete

File: alv-portal-ui/src/app/shared/forms/input/file-input/drag-drop.directive.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  EventEmitter,
+  HostBinding,
+  HostListener, Input,
+  Output
+} from '@angular/core';
+
+@Directive({
+  selector: '[alvDragDrop]'
+})
+export class DragDropDirective {
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/forms/input/file-input/drag-drop.directive.ts]: rename to FileDragDropDirective

File: alv-portal-ui/src/app/shared/forms/input/file-input/file/file.component.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import { Observable } from 'rxjs';
+import { FileSaverService } from '../../../../file-saver/file-saver.service';
+import { FileMetadata } from './file.types';
+
+@Component({
+  selector: 'alv-file',
+  templateUrl: './file.component.html',
+  styleUrls: ['./file.component.scss']
+})
+export class FileComponent implements OnInit {
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/forms/input/file-input/file/file.component.ts]: rename to FileUploadPreviewComponent

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    this.control.markAsTouched();
+
+    if (!this.showRemoveButton) {
+      this.files = []; // reset because there is no delete button
+    }
+
+    let fileTooLarge = false;
+    let maxFilesCount = false;
+    let invalidFileType = false;
+    for (let i = 0; i < files.length; i++) {
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts]: High Complecity -> can we reduce that or write unit tests for each break and continue case using mutiple files

File: alv-portal-ui/src/app/shared/pipes/markdown-escape.pipe.ts [hunk 11]
```diff
@@ -22,7 +22,7 @@ export class MarkdownEscapePipe implements PipeTransform {
       // Remove horizontal rules
         .replace(/^(-\s*?|\*\s*?|_\s*?){3,}\s*$/gm, '')
         // Strip list leaders
-        .replace(/^([\s\t]*)([\*\-\+]|\d+\.)\s+/gm, '$1')
+        .replace(/^([\s\t]*)([\*\-\+•]|\d+\.)\s+/gm, '$1')
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/pipes/markdown-escape.pipe.ts]: ???

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.component.ts (hunk 2 above)
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.component.ts]: done

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.model.ts (hunk 3 above)
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-document/application-document.model.ts]: done

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts (hunk 4 above)
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: done

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-sorting.service.ts (hunk 6 above)
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-sorting.service.ts]: done

File: alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.repository.ts (hunk 7 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.repository.ts]: no, it's being used by findByOwnerUserId

File: alv-portal-ui/src/app/shared/forms/input/file-input/drag-drop.directive.ts (hunk 8 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/file-input/drag-drop.directive.ts]: done

File: alv-portal-ui/src/app/shared/forms/input/file-input/file/file.component.ts (hunk 9 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/file-input/file/file.component.ts]: done

File: alv-portal-ui/src/app/shared/pipes/markdown-escape.pipe.ts (hunk 11 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/pipes/markdown-escape.pipe.ts]: removed

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts (hunk 10 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts]: There are already some unit tests in `file-input.component.spec.ts`

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts (hunk 5 above)
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: I have split up the calculation in two methods.. and I use reduce now

File: alv-portal-ui/src/app/showcase/showcase.component.ts [hunk 12]
```diff
@@ -99,6 +102,9 @@ export class ShowcaseComponent implements OnInit {
       });
   }
 
+  logFiles(files: File[]) {
+    console.log(files);
```
Comment (pavax) [inline alv-portal-ui/src/app/showcase/showcase.component.ts]: remove logging

Comment (99999989) [reply alv-portal-ui/src/app/showcase/showcase.component.ts]: It's in the showcase.component .. just to show there is something

File: alv-portal-ui/src/app/shared/forms/input/select/select.component.html [hunk 13]
```diff
@@ -15,6 +15,7 @@
             [attr.aria-required]="!!required"
             [attr.aria-describedby]="validationId"
             [alvAutofocus]="autofocus"
+            (change)="control.markAsTouched()"
```
Comment (pavax) [inline alv-portal-ui/src/app/shared/forms/input/select/select.component.html]: how is this change related to that pr?

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts (hunk 10 above)
Comment (pavax) [reply alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts]: is there also a unit test for multiple files?

File: alv-portal-ui/src/app/shared/forms/input/select/select.component.html (hunk 13 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/select/select.component.html]: It's an old bug we didn't notice before, because in 99% of the cases we preselect a valid value in the select.component. Without this change, the select.component never gets invalid (red) because it's not touched.

Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/select/select.component.html]: And yes, it's related to this PR :)

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+  applySortByDocumentType() {
+    this.applicationDocuments = this.applicationDocumentsSortingService.sortByDocumentType(this.applicationDocuments);
+  }
+
+  applySortByFilename() {
+    this.applicationDocuments = this.applicationDocumentsSortingService.sortByFilename(this.applicationDocuments);
+  }
+
+  private calculateInvalidDocumentTypes(excludedDocumentType: ApplicationDocumentType): ApplicationDocumentType[] {
+    return Object.values(ApplicationDocumentType).reduce((invalidDocumentTypes, currentDocumentType) => {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: to me it seems more like a map than a reduce: we turn the array `Object.values(ApplicationDocumentType)` into an array `invalidDocumentTypes`.

File: alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.types.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+export interface ApplicationDocument {
+  id?: string;
+  documentType: ApplicationDocumentType;
+  ownerUserId: string;
+  createdAt: string;
+  updatedAt: string;
+  documentMetadata: DocumentMetadata;
+}
+
+export interface CreateApplicationDocument {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.types.ts]: 'create' is a verb, we only name functions with the verbs

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+      if (this.maxFileSize && files[i].size > this.maxFileSize) {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts]: I would extract all of those if's into a functions

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-upload-preview/file-upload-preview.component.ts [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+  getFileIcon(fileType: string): string[] {
+    const fileIcon = this.iconMappings[fileType];
+    if (fileIcon) {
+      return fileIcon;
+    }
+    return ['far', 'file'];
+  }
+
+  downloadFile() {
```
Comment (ganqqwerty) [inline alv-portal-ui/src/app/shared/forms/input/file-input/file-upload-preview/file-upload-preview.component.ts]: here it looks suspicious. Subscribe happens on click? What will happen if I click multiple times? Will it be debounced, or will my next clicks will return me the cached version of the file?

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+  getAllApplicationDocuments() {
+    this.authenticationService.getCurrentUser().pipe(
+      filter(user => !!user),
+      flatMap(user => this.applicationDocumentsRepository.findByOwnerUserId(user.id)),
+    ).subscribe(applicationDocuments => {
+      this.applicationDocuments = this.applicationDocumentsSortingService.sort(applicationDocuments, this.selectedSortType);
+    });
+  }
+
+  openModal(applicationDocument?: ApplicationDocument) {
```
[the remark this answers is not in the document]
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: This problem exists with every modal we use this way. But you are right, it's annoying. I can increase the z-index for the notifications so that they are shown on top of the modal backdrop.

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+  applySortByDocumentType() {
+    this.applicationDocuments = this.applicationDocumentsSortingService.sortByDocumentType(this.applicationDocuments);
+  }
+
+  applySortByFilename() {
+    this.applicationDocuments = this.applicationDocumentsSortingService.sortByFilename(this.applicationDocuments);
+  }
+
+  private calculateInvalidDocumentTypes(excludedDocumentType: ApplicationDocumentType): ApplicationDocumentType[] {
```
[the remark this answers is not in the document]
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: excluded from the calculation of invalid document types. It's for the edit case: the already existing document type can always be selected

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts (hunk 14 above)
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-overview.component.ts]: It was a `filter` before, but @pavax said it looks more like a `reduce`, so I changed it.

File: alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-sorting.service.ts [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        return this.sortByDate(applicationDocuments);
+      case ApplicationDocumentSortType.BY_DOCUMENT_TYPE:
+        return this.sortByDocumentType(applicationDocuments);
+      case ApplicationDocumentSortType.BY_FILENAME:
+        return this.sortByFilename(applicationDocuments);
+    }
+  }
+
+  private dateComparator(a: ApplicationDocument, b: ApplicationDocument): number {
+    return new Date(a.updatedAt).getTime() - new Date(b.updatedAt).getTime();
```
[the remark this answers is not in the document]
Comment (99999989) [reply alv-portal-ui/src/app/online-services/application-documents/application-documents-overview/application-documents-sorting.service.ts]: The dateComparator needs to return a `number` (1, 0 or -1), not a `boolean`

File: alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.types.ts (hunk 15 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.types.ts]: Screenshot 2019-07-29 at 20 51 18

Actually we used this kind of naming many times already. I just wanted to stick with the pattern.

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts (hunk 10 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts]: Yes, it's called `should reject too many files and disable button`

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts (hunk 16 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/file-input/file-input.component.ts]: done

File: alv-portal-ui/src/app/shared/forms/input/file-input/file-upload-preview/file-upload-preview.component.ts (hunk 17 above)
Comment (99999989) [reply alv-portal-ui/src/app/shared/forms/input/file-input/file-upload-preview/file-upload-preview.component.ts]: A subscribe should not happen on a click event? It's just a GET request that would be executed twice.. I don't see why that's bad. Actually, Chrome even blocks the double file download.

File: alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.types.ts (hunk 15 above)
Comment (ganqqwerty) [reply alv-portal-ui/src/app/shared/backend-services/application-documents/application-documents.types.ts]: ah, ok
